sky/skylet/providers/ibm/node_provider.py

```diff
@@ -0,0 +1,924 @@
+        """
+        Args:
+            provider_config (dict): containing the provider segment of
+                the cluster's config file.initialized as an instance
+                variable of the parent class, hence can accessed via self.
+            cluster_name(str): value of cluster_name within the
+                 cluster's config file.
+
+        """
+        NodeProvider.__init__(self, provider_config, cluster_name)
+
+        self.lock = threading.RLock()
+        self.tags_file = Path.home() / VPC_TAGS
+        self.iam_api_key = provider_config["iam_api_key"]
+        self.resource_group_id = provider_config["resource_group_id"]
+        self.iam_endpoint = provider_config.get("iam_endpoint")
+        self.region = provider_config["region"]
+        self.endpoint = f"https://{self.region}.iaas.cloud.ibm.com"
+        self.zone = self.provider_config["availability_zone"]
+        self.ibm_vpc_client = _get_vpc_client(
+            self.endpoint, IAMAuthenticator(
+                self.iam_api_key, url=self.iam_endpoint)
+        )
+        self.vpc_tags = None
+        self.vpc_provider = IBMVPCProvider(
+            self.resource_group_id, self.region, self.cluster_name)
+
+        self._load_tags()
+
+        # Cache of starting/running/pending(below PENDING_TIMEOUT)nodes.
+        #                               in the format{node_id:node_data}.
+        self.cached_nodes = {}
+        # cache of the nodes created, but not yet tagged and running.
+        #                    in the format {node_id:time_of_creation}.
+        self.pending_nodes = {}
+        self.deleted_nodes = []  # ids of nodes scheduled for deletion.
+
+        # if cache_stopped_nodes == true, nodes will be stopped
+        #   instead of deleted to better accommodate future rise in demand
+        self.cache_stopped_nodes = provider_config.get(
+            "cache_stopped_nodes", True)
+
+    def _get_node_type(self, name):
+        if f"{self.cluster_name}-{NODE_KIND_WORKER}" in name:
+            return NODE_KIND_WORKER
+        elif f"{self.cluster_name}-{NODE_KIND_HEAD}" in name:
+            return NODE_KIND_HEAD
+
+    def _get_nodes_by_tags(self, filters):
+        """
+        returns list of nodes who's tags are matching the specified filters. 
+        Args:
+            filters(dict):  specified conditions to filter nodes by.
+        """
+
+        nodes = []
+        # no filters were specified or the only filter is the type of the node
+        if (not filters or
+                list(filters.keys()) == [TAG_RAY_NODE_KIND]):
+            result = self.ibm_vpc_client.list_instances().get_result()
+            instances = result["instances"]
+            while result.get("next"):
+                start = result["next"]["href"].split("start=")[1]
+                result = self.ibm_vpc_client.list_instances(
+                    start=start).get_result()
+                instances.extend(result["instances"])
```

Comment:
The row `kind = self._get_node_type(instance["name"])` already takes care of that, by looking not only for the TYPE of node but also for the CLUSTER_NAME. i've added it to the function's documentation, it does require explicit acknowledgment.

As i wrote in my previous messages, now that vpc_provider is implemented and tagged, guaranteeing all instances within it belonging to the cluster, it would be much more efficient to collect cluster nodes by adding `vpc_id` as a filtering parameter to `get_instances()`. 
I could try to implement it in this PR, which will further postpone it, or i could add it to my todo list (along with --image-id support for example).

Reply:
Sounds good!